### DISCUSSION TITLE
Attempted reference resolution before document reached ComputedScopes state

### DISCUSSION THREAD

**Discussion by ejfasting** [eclipse-langium/langium#1523] [Q&A] [answered]
Hello, 

I'm using langium as one of many embedded languages. It's beging used like this:
```
const langiumDocument = sharedService.workspace.LangiumDocumentFactory.fromTextDocument(document);
const params = { textDocument: document, position: position };
const hoverContent = await definitionService.lsp.HoverProvider?.getHoverContent(langiumDocument, params, token);
```
In my case what i hover is a cross reference. 
This results in the error: Attempted reference resolution before document reached ComputedScopes state.

From testing it looks like the `startLanguageServer`, which starts up the 'native langium server', contains the method `addDocumentUpdateHandler` . This looks to be mandatory to get the hover resolve properly. Unfortunately this brings other issues for my LSP server, and it actually breaks the Diagnostics (`validation.DocumentValidator.validateDocument` no longer provides any diagnostics). So i was hoping to avoid adding things specifically for langium to the whole LSP server.

Is it supposed to be possible to 'only' pass a langiumDocument to a provider and have it parse the content directly? Or do you have to add a DocumentUpdateHandler for it to work properly?

//Eivind

**Comment by msujew** [comment, marked as answer]
Hey @ejfasting,

in order to provide hovers for targets of cross references, i.e.:

```ts
const value = 42;
value // <-- hover here
```

Langium needs to resolve those cross references to figure out what kind of information to provide on hover. Generally, the `DocumentBuilder` should be invoked to build documents. Langium is simply not designed to provide LSP results without first going through the document lifecycle.

Ideally, you should be able to replicate the document handling that is happening inside of `addDocumentUpdateHandler`. I've recommended this also already in https://github.com/eclipse-langium/langium/discussions/1494#discussioncomment-9433725, as these services are meant to cache relevant data across rebuilds so that you don't run into performance issues in your language server. Reparsing and relinking a document on every hover/go-to/completion request is pretty inefficient. Just calling `DocumentBuilder#update` should be enough to handle almost everything.
